internal/cli/console.go

```diff
@@ -0,0 +1,236 @@
+	GlobalOptions
+}
+
+func DefaultConsoleOptions() *ConsoleOptions {
+	return &ConsoleOptions{
+		GlobalOptions: DefaultGlobalOptions(),
+	}
+}
+
+func NewConsoleCmd() *cobra.Command {
+	o := DefaultConsoleOptions()
+
+	cmd := &cobra.Command{
+		Use:   "console device/NAME",
+		Short: "Connect a console to the remote device through the server.",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.Complete(cmd, args); err != nil {
+				return err
+			}
+			if err := o.Validate(args); err != nil {
+				return err
+			}
+			return o.Run(cmd.Context(), args)
+		},
+		SilenceUsage: true,
+	}
+
+	o.Bind(cmd.Flags())
+
+	return cmd
+}
+
+func (o *ConsoleOptions) Bind(fs *pflag.FlagSet) {
+	o.GlobalOptions.Bind(fs)
+}
+
+func (o *ConsoleOptions) Complete(cmd *cobra.Command, args []string) error {
+	return o.GlobalOptions.Complete(cmd, args)
+}
+
+func (o *ConsoleOptions) Validate(args []string) error {
+	kind, name, err := parseAndValidateKindName(args[0])
+	if err != nil {
+		return err
+	}
+
+	if kind != DeviceKind {
+		return fmt.Errorf("only devices can be connected to a console")
+	}
+
+	if len(name) == 0 {
+		return fmt.Errorf("device name is required")
+	}
+	return nil
+}
+
+func (o *ConsoleOptions) Run(ctx context.Context, args []string) error { // nolint: gocyclo
+	c, err := client.NewFromConfigFile(o.ConfigFilePath)
+	if err != nil {
+		return fmt.Errorf("creating client: %w", err)
+	}
+
+	_, name, err := parseAndValidateKindName(args[0])
+	if err != nil {
+		return err
+	}
+	console, err := c.RequestConsoleWithResponse(ctx, name)
+
+	if err != nil {
+		return fmt.Errorf("error requesting console: %w", err)
+	}
+
+	if console.HTTPResponse.StatusCode != 200 {
+		return fmt.Errorf("error requesting console: %s, %+v", console.HTTPResponse.Status, console.HTTPResponse.Body)
+	}
+
+	grpcEndpoint := console.JSON200.GRPCEndpoint
+	sessionID := console.JSON200.SessionID
+
+	err = o.connectViaGRPC(ctx, grpcEndpoint, sessionID)
+	if err == io.EOF {
+		fmt.Println("Connection closed")
+		return nil
+	}
+	return err
+}
+
+// TODO: Move this to a websocket call instead later, the console endpoint will redirect to a ws method
+func (o *ConsoleOptions) connectViaGRPC(ctx context.Context, grpcEndpoint, sessionID string) error {
+	//grpcEndpoint = "grpcs://192.168.1.10:7444"
+	grpcEndpoint = strings.TrimRight(grpcEndpoint, "/")
+	fmt.Printf("Connecting to %s with session id %s\n", grpcEndpoint, sessionID)
+	client, err := client.NewGrpcClientFromConfigFile(o.ConfigFilePath, grpcEndpoint)
+	if err != nil {
+		return fmt.Errorf("creating grpc client: %w", err)
+	}
+	// add key-value pairs of metadata to context
+	ctx = metadata.AppendToOutgoingContext(ctx, agentserver.SessionIDKey, sessionID)
+	ctx = metadata.AppendToOutgoingContext(ctx, agentserver.ClientNameKey, "flightctl-cli")
+
+	stream, err := client.Stream(ctx)
+	if err != nil {
+		return fmt.Errorf("error creating stream: %w", err)
```

Comment:
If there is a failure to connect, shouldn't the server be notified (by REST) to free the session id?

Reply:
I believe we should just rely on timeouts to keep things simple.